### ISSUE TITLE
Contract deployments with `blockTime` set take ~4 seconds

### ISSUE THREAD

**Issue by ndzik** [ConsenSys-archive/ganache#1540]
TLDR:
[Minimal working example (MWE)](https://github.com/ndzik/funwithganache)
Using the the current `ganache@alpha` npm package together with the latest version of `ethersjs` exhibits an unexpected behavior when deploying contracts. The deployment of a single contract takes ~4 seconds **unless** a seemingly random but once found consistent amount of blocks is force mined with the `evm_mine` call.

For some production code I needed to setup a blockchain environment which has some contracts predeployed for later uses. We were using `ethereum-waffle@3.4.0`, which under the hood seems to use `ganache-core@2.13.2`. Some tests are time dependent which required me to allow the use of the `blockTime` option for the underlying provider. Once this `blockTime` was set, the setup for the tests took a long time, always around ~4+ seconds per deployed contract. I tried speeding them up by triggering the mining of blocks with the previously mentioned `evm_mine` call, but this did not help. Playing around even further showed that in that instance mining 7 blocks immediately triggered the contract deployment.

So I created a [MWE](https://github.com/ndzik/funwithganache) with `ganache@alpha` and an `ethersjs@5.5.1` which shows this behavior and logs the provider output to corresponding files. It already contains logs for my run and the console output of the timings is the following:

```
Time taken without blockTime set: 4285.957789000124
Time taken with 1 skipped: 4176.430750001222
Time taken with 12 skipped: 4139.068713000044
Time taken with 13 skipped: 4129.8051149994135
Time taken with 18 skipped: 149.827926998958
Time taken with 19 skipped: 90.71113900095224
```

Interestingly using `ganache@alpha` with **no** `blockTime` now also takes 4+ seconds, but this might just be a misconfiguration on my end?

Notes: I am fully aware that `ethersjs` sets a default polling rate of 4 seconds but I cannot believe that this is connected, since in my tests setting the `pollingRate` did not change my time measurements **and** deploying a contract is not event based. Unless there is peculiar stuff going on behind the curtains which I am not aware of. But then this brings up another question: Why does skipping `n` amount of blocks suddenly trigger the deployment if this is related to `ethersjs` polling rate?

I hope I can get some help and that the [MWE](https://github.com/ndzik/funwithganache) is enough. To run simply issue `yarn && yarn start` and the logs will be put under the `<project>/logs/` directory. And the MWE can be adapted as seen fit by whoever this might concern.

Thanks for the help in advance :pray:

**Comment by davidmurdoch** [comment]
Thanks for that MWE! It was very helpful in debugging this.

Keep in mind that most everything here is happening asynchronously. So while your code `await`s in one place, this won't prevent other microtasks (V8's name for deferred tasks, like awaited Promises) from being executed.

What is happening in the 4+ second cases:

1. the deployment transaction _request_ is sent
2. in the meantime, `evm_mine` is called a few times, maybe for the last time
3. the deployment transaction request is received by ganache, but not yet processed
4. in the meantime, `evm_mine` may be called a few more times, maybe for the last time
5. the received transaction is processed and added to the transaction pool by Ganache
6. ether.js finally receives the transaction hash from Ganache then immediately calls `eth_getTransactionReceipt`; it returns `null`.
7. Ganache mines the transaction as scheduled by `blockTime`
8. 4 seconds later ethers.js polls `eth_getTransactionReceipt` and finds the transaction

And in the under 4 second cases:

1. the deployment transaction _request_ is sent
2. in the meantime, `evm_mine` is called a few times
3. the deployment transaction request is received by ganache, but not yet processed
4. in the meantime, `evm_mine` is called a few times
5. the received transaction is processed and added to the transaction pool by Ganache
6. **`evm_mine` is called once again and _immediately_ executes all pending transactions _ahead_ of the scheduled `blockTime` operation.**
   * this is intended behavior
7. ethers.js finally receives the transaction hash from Ganache then immediately calls `eth_getTransactionReceipt` and finds the receipt.

---

If you add `provider.pollingInterval = 1; // 1 millisecond` after initializing the provider the runtimes will be reduced to about 1 second.

---


What I'd love to see from ether.js here is use of `eth_subscribe` for `newHeads`. That way it would resolve the contract deployment without polling. I believe this capability is built in to ethers.js already, but the Ganache provider may not be compatible with the subscription mechanism or the timing may be off. @ricmoo Do you have an ideas on how a ganache provider can opt in to this subscriptions, if possible (FYI: this is all on an [alpha version](https://github.com/trufflesuite/ganache/releases/tag/ganache%407.0.0-alpha.1) of ganache v7)?

**Comment by davidmurdoch** [comment]
Thanks for that MWE! It was very helpful in debugging this.

Keep in mind that most everything here is happening asynchronously. So while your code `await`s in one place, this won't prevent other microtasks (V8's name for deferred tasks, like awaited Promises) from being executed.

What is happening in the 4+ second cases:

1. the deployment transaction _request_ is sent
2. in the meantime, `evm_mine` is called a few times, maybe for the last time
3. the deployment transaction request is received by ganache, but not yet processed
4. in the meantime, `evm_mine` may be called a few more times, maybe for the last time
5. the received transaction is processed and added to the transaction pool by Ganache
6. ether.js finally receives the transaction hash from Ganache then immediately calls `eth_getTransactionReceipt`; it returns `null`.
7. Ganache mines the transaction as scheduled by `blockTime`
8. 4 seconds later ethers.js polls `eth_getTransactionReceipt` and finds the transaction

And in the under 4 second cases:

1. the deployment transaction _request_ is sent
2. in the meantime, `evm_mine` is called a few times
3. the deployment transaction request is received by ganache, but not yet processed
4. in the meantime, `evm_mine` is called a few times
5. the received transaction is processed and added to the transaction pool by Ganache
6. **`evm_mine` is called once again and _immediately_ executes all pending transactions _ahead_ of the scheduled `blockTime` operation.**
   * this is intended behavior
7. ethers.js finally receives the transaction hash from Ganache then immediately calls `eth_getTransactionReceipt` and finds the receipt.

---

If you add `provider.pollingInterval = 1; // 1 millisecond` after initializing the provider the runtimes will be reduced to about 1 second.

---


What I'd love to see from ether.js here is use of `eth_subscribe` for `newHeads`. That way it would resolve the contract deployment without polling. I believe this capability is built in to ethers.js already, but the Ganache provider may not be compatible with the subscription mechanism or the timing may be off. @ricmoo Do you have an ideas on how a ganache provider can opt in to this subscriptions, if possible (FYI: this is all on an [alpha version](https://github.com/trufflesuite/ganache/releases/tag/ganache%407.0.0-alpha.1) of ganache v7)?

**Comment by ricmoo** [comment]
In v5, the getChanges functions are not used in any of the polling providers, since it conflicts with more complex providers, like the FallbackProvider.

I’m working on that part right now for v6, where there is a Subscriber object, which can use polling, subscribe with getChanges or rely on multiplexing duplex channels (like WebSockets, IPC, future things).

For v5, if you use a WebSocket you will get the semantics you are expecting though, so that’s the quickest stopgap until v6 is out.

Does that make sense?

**Comment by ricmoo** [comment]
Oh, also for ganache you should probably also use `provider.pollingInterval = 100;` to reduce the amount of time it waits. Won’t help you with the ordering, but will make things speedier.

**Comment by davidmurdoch** [comment]
@ricmoo yes, thank you. That makes sense.

Not sure if ethers.js already handles this today for the providers that do use the getChanges functions, but something to keep in mind with Ganache v7's alpha's new default "mining" mode is that it will now return the transaction hash _before_ the transaction is mined, and will then mine it immediately. Previously ganache would wait to return the hash until the transaction was "confirmed" locally. So if ethers.js waits to subscribe to changes until _after_ the transaction is sent there is a good chance it will miss the transaction's block's `newHeads` event (depending on latency and/or I/O).

Example: 

This may not subscribe in time:

```javascript
const txHash = await provider.request({method: "eth_sendTransaction", params:[tx]});
// it _might_ be too late to subscribe here
await provider.request({method: "eth_subscribe", params:["newHeads"]});
provider.on("message", checkForReceipt);
```

whereas this would:
```javascript
await provider.request({method: "eth_subscribe", params:["newHeads"]});
const txHash = await provider.request({method: "eth_sendTransaction", params:[tx]});
// note: listening for `message` event immediately after receipt of the txHash is okay
provider.on("message", checkForReceipt);

```

**Comment by ndzik** [comment]
Thanks a lot for the detailed answers and clarifications :rocket:. So just to be sure: Did I understand correctly that switching to a `WebSocketProvider`, if possible, will yield the expected result?

I will try that later today and respond here again!

**Comment by ricmoo** [comment]
@ndzik As best as I would guess, yes, a WebSocketProvider should work. :)

**Comment by MicaiahReid** [comment]
Closing for issue maintenance; it doesn't look like there's anything for us to do here.

**Comment by ndzik** [comment]
Sorry for the late reply, I got sick shortly after opening this issue and did not come around to it again until today.

Thanks again for all the replies and I can confirm that using `WebSocketProvider` behaves as expected :muscle:
